test/server/guarddog_impl_test.cc

```diff
@@ -355,6 +438,360 @@ TEST_P(GuardDogTestBase, AtomicIsAtomicTest) {
   ASSERT_EQ(atomic_time.is_lock_free(), true);
 }
 
+// A GuardDogAction used for testing the GuardDog.
+// It's primary use is dumping string of the format EVENT_TYPE : tid1,.., tidN to
+// the events vector passed to it.
+// Instances of this class will be registered for GuardDogEvent through
+// TestGuardDogActionFactory.
+class LogGuardDogAction : public Configuration::GuardDogAction {
+public:
+  LogGuardDogAction(std::vector<std::string>& events) : events_(events) {}
+
+  void run(envoy::config::bootstrap::v3::Watchdog::WatchdogAction::WatchdogEvent event,
+           std::vector<std::pair<Thread::ThreadId, MonotonicTime>> thread_ltt_pairs,
+           MonotonicTime /*now*/) override {
+    std::string event_string =
+        envoy::config::bootstrap::v3::Watchdog::WatchdogAction::WatchdogEvent_Name(event);
+    absl::StrAppend(&event_string, " : ");
+    std::vector<std::string> output_string_parts;
+
+    for (const auto& thread_ltt_pair : thread_ltt_pairs) {
+      output_string_parts.push_back(thread_ltt_pair.first.debugString());
+    }
+
+    absl::StrAppend(&event_string, absl::StrJoin(output_string_parts, ","));
+    events_.push_back(event_string);
+  }
+
+protected:
+  std::vector<std::string>& events_; // not owned
+};
+
+// A GuardDogAction that raises the specified signal.
+class SignalGuardDogAction : public Configuration::GuardDogAction {
+public:
+  SignalGuardDogAction(int signal) : signal_(signal) {}
+
+  void run(envoy::config::bootstrap::v3::Watchdog::WatchdogAction::WatchdogEvent /*event*/,
+           std::vector<std::pair<Thread::ThreadId, MonotonicTime>> /*thread_ltt_pairs*/,
+           MonotonicTime /*now*/) override {
+    std::raise(signal_);
+  }
+
+protected:
+  const int signal_;
+};
+
+// Test factory for consuming Watchdog configs and creating GuardDogActions.
+template <class ConfigType>
+class TestGuardDogActionFactory : public Configuration::GuardDogActionFactory {
+public:
+  TestGuardDogActionFactory(const std::string& name, std::vector<std::string>& events)
+      : name_(name), events_(events) {}
+
+  Configuration::GuardDogActionPtr createGuardDogActionFromProto(
+      const envoy::config::bootstrap::v3::Watchdog::WatchdogAction& config,
+      Configuration::GuardDogActionFactoryContext& /*context*/) override {
+    // Return different actions depending on the config.
+    if (config.event() == envoy::config::bootstrap::v3::Watchdog::WatchdogAction::KILL) {
+      return std::make_unique<SignalGuardDogAction>(SIGKILL);
```

Comment:
That's a great idea! 

I needed some signal that would kill the proc, and I needed to differentiate it from the `PANIC(..)` call that would occur in kill / multikill after the handlers ran. Since the `PANIC(..)` was a `SIGABRT` ultimately, I played around with using different signals, but your idea is more on target, thanks.